core/src/main/java/org/apache/calcite/plan/volcano/VolcanoPlanner.java

```diff
@@ -167,7 +137,13 @@
   /**
    * Listener for this planner, or null if none set.
    */
-  RelOptListener listener;
+  MulticastRelOptListener listener;
+
+  /**
+   * Listener for couting the attempts of each rule.
+   * Only enabled under DEBUG.
+   */
+  RuleAttemptsListener ruleAttemptsListener;
```

Comment:
Is there a special reason we add this as a class member? 
A planner may have a number of different listeners.

Reply:
The reason is you would need to access it at the end to dump rule information.

Reply:
Seems not necessary, either we get the `RuleAttemptsListener ` out of the `MulticastRelOptListener` or we add a new interface `RelOptListener#dump` or `RelOptListener#close` .

Reply:
Getting RuleAttemptsListener out of MulticastRelOptListener would require a down casting and relay on the implementation detail that the RuleAttemptsListener is always the first listener, which is tricky. And the dump() method is specific to the RuleAttemptsListener, and I don't think it should become part of the interface.